### PR TITLE
T/116

### PR DESCRIPTION
**Pull request by ma2ciek**
Resolves https://github.com/ckeditor/ckeditor5-utils/issues/116.

Translation service can't use 'define' function because of the https://github.com/ckeditor/ckeditor5-dev/issues/59 issue.

### CODE REVIEW

File: src/locale.js [hunk 1]
```diff
@@ -53,10 +55,14 @@ export default class Locale {
 	 * @private
 	 */
 	_t( str, values ) {
+		let trasnaltedString = translate( this.lang, str );
```
Comment (Reinmar) [inline src/locale.js]: Typo.

File: src/translation-service.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+/**
+ * @module utils/translation-service
+ */
+
+const translations = {};
+
+/**
+ * Merges package translations to existing ones.
```
Comment (Reinmar) [inline src/translation-service.js]: It doesn't merge anything – it simply adds translations. Do not overcomplicate docs ;).

File: src/translation-service.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * These translations can be used later with {@link module:utils/translations-service~translate translate}
```
Comment (Reinmar) [inline src/translation-service.js]: Missing period.

File: src/translation-service.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ *		add( 'pl', {
+ *			'OK': 'OK',
+ *			'Cancel [context: reject]': 'Anuluj'
+ *		} );OK
```
Comment (Reinmar) [inline src/translation-service.js]: Typo.

File: src/translation-service.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @param {String} lang
```
Comment (Reinmar) [inline src/translation-service.js]: Missing param description

File: src/translation-service.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * @param {Object.<String, Object>} packageDictionary
```
Comment (Reinmar) [inline src/translation-service.js]: You don't know if this is a package's dictionary. These are translations and can come from any source.

This option also misses description and the type is wrong too (values are not objects).

File: src/translation-service.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * @returns undefined
```
Comment (Reinmar) [inline src/translation-service.js]: No need to document that;

File: src/translation-service.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+export function add( lang, packageDictionary ) {
+	if ( !( lang in translations ) ) {
+		translations[ lang ] = {};
+	}
+
+	const dictionary = translations[ lang ];
+
+	for ( const translationKey in packageDictionary ) {
+		const translation = packageDictionary[ translationKey ];
+		dictionary[ translationKey ] = translation;
```
Comment (Reinmar) [inline src/translation-service.js]: Overcomplicated. Please merge these two lines.

Besides, `Object.assign()` should be better.

File: tests/translation-service.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import { translate, add, _clear } from '../src/translation-service';
+
+describe( 'translationService', () => {
```
Comment (Reinmar) [inline tests/translation-service.js]: Wrong name – it's `translation-service`.

File: src/translation-service.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+/**
+ * Translates string if the translation of the string was previously added using {@link module:utils/translations-service~add add}.
+ * Otherwise returns original (English) sentence.
+ *
+ *		translate( 'pl', 'Cancel [context: reject]' );
+ *
+ * @param {String} lang Translation language.
+ * @param {String} translationKey Sentence which is going to be translated.
```
Comment (Reinmar) [inline src/translation-service.js]: It doesn't have to be a sentence.

File: src/translation-service.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+	}
+}
+
+/**
+ * Translates string if the translation of the string was previously added using {@link module:utils/translations-service~add add}.
+ * Otherwise returns original (English) sentence.
+ *
+ *		translate( 'pl', 'Cancel [context: reject]' );
+ *
+ * @param {String} lang Translation language.
```
Comment (Reinmar) [inline src/translation-service.js]: Incorrect description. It's not "translation's language" (it doesn't make any sense). It's more like "Target language." or "Language which dictionary will be used.".

File: src/translation-service.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ */
+export function translate( lang, translationKey ) {
+	if ( !existTranslationKey( lang, translationKey ) ) {
+		return translationKey.replace( / \[context: [^\]]+\]$/, '' ) ;
+	}
+
+	return translations[ lang ][ translationKey ];
+}
+
+function existTranslationKey( lang, translationKey ) {
```
Comment (Reinmar) [inline src/translation-service.js]: `hasTranslation()`

Comment (Reinmar) [inline src/translation-service.js]: or `hasTranslationFor()`

File: src/translation-service.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ *		translate( 'pl', 'Cancel [context: reject]' );
+ *
+ * @param {String} lang Translation language.
+ * @param {String} translationKey Sentence which is going to be translated.
+ * @returns {String} Translated sentence.
+ */
+export function translate( lang, translationKey ) {
+	if ( !existTranslationKey( lang, translationKey ) ) {
+		return translationKey.replace( / \[context: [^\]]+\]$/, '' ) ;
```
Comment (Reinmar) [inline src/translation-service.js]: Typo.

File: src/translation-service.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ * Otherwise returns original (English) sentence.
+ *
+ *		translate( 'pl', 'Cancel [context: reject]' );
+ *
+ * @param {String} lang Translation language.
+ * @param {String} translationKey Sentence which is going to be translated.
+ * @returns {String} Translated sentence.
+ */
+export function translate( lang, translationKey ) {
+	if ( !existTranslationKey( lang, translationKey ) ) {
```
Comment (Reinmar) [inline src/translation-service.js]: I'm not sure whether this code makes any sense (and if it does, it requires comments). Please explain me how's it going to work in the dev env (where no translations are defined and `t()` is called with the original strings) and in the situation when translation keys were replaced with chosen language. AFAICT, it's going to work in both of these situations, but I'd like to have confirmation and comments for that. Perhaps the best way for them will be the documentation of this function which now is incorrect ("returns original (English) sentence" is not entirely true in the second mentioned situation).

Please also remember to explain the not yet implemented support for multiple languages.

File: src/translation-service.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+	);
+}
+
+/**
+ * Clears translations for test purpose.
+ *
+ * @protected
+ */
+export function _clear() {
+	for ( const key in translations ) {
```
Comment (Reinmar) [inline src/translation-service.js]: `translations = {};` would be easier.

File: src/translation-service.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module utils/translation-service
+ */
+
+const translations = {};
```
Comment (Reinmar) [inline src/translation-service.js]: Maybe `dictionaries`? You later call values of this object `dictionary` (L30).

File: src/translation-service.js (hunk 8 above)
Comment (ma2ciek) [inline src/translation-service.js]: Ow, I missed that after the concept has changed.

**Pull request merged**